### ISSUE TITLE
Swapping the order of episodes on lists and loops

### ISSUE THREAD

**Issue by maxim-belkin** [swcarpentry/python-novice-inflammation#496]
It has been decided (see #453) that **lists** should be introduced _before_ **loops**.

Because this change might take some time to implement, it was decided to first pursue this effort in a separate branch and later merge this branch into the main `gh-pages` branch.

This special branch has been created and is called `lists-loops`. We, therefore, ask anyone who is interested in contributing to this effort to submit Pull Requests against this `lists-loops` branch.

### How to contribute

Start by fetching and checking out the `lists-loops` branch into your current repository

```bash
git fetch https://github.com/swcarpentry/python-novice-inflammation.git lists-loops
git checkout lists-loops
```

Now, you can start working on it as usual:
```bash
git checkout -b my-changes lists-loops
# ... make changes...
git add -u
git commit
git push <your-fork>
```
... and submit a PR! When you submit a pull request related to swapping the order of episodes on loops and lists, make sure to select `lists-loops` branch of this repository as the base branch.

That's it!

_We hope to see **your** Pull Request soon_!

**Comment by bhickson** [comment]
@maxim-belkin 
For a carpentries lesson this last February I did some reorg and partial rewrite of the python lesson into Jupyter notebooks that follows this idea. Basically introducing variables, data types, structures,  statements, conditionals, loops, etc in that order. Would this order be worth considering for this issue/branch?

https://github.com/UA-Carpentries-Workshops/2018-02-10-Tucson/blob/master/python_lessons/01%20-%20Introduction%20to%20Programming%20with%20Python.ipynb

If so, I'd be happy bring them back out of the notebook to the branch.

**Comment by maxim-belkin** [comment]
@bhickson, thanks for offering your help! The notebook you reference has changes that might be appropriate for both branches (`gh-pages` and `lists-loops`) so please feel free to suggest your changes _via_ pull requests.

**Comment by kekoziar** [comment]
Is there anything else that needs to be worked on for this branch?

**Comment by biokcb** [comment]
Hi, I looked over the changes in these lessons a bit. I agree that it makes sense to introduce lists before loops and the transition to looping is a lot clearer now. However, around line 195 in `03-loop.md` it jumps straight into iterating over the characters in a string again, without mentioning that characters in a string can be accessed by indexing like we've done with lists. The old lesson had this in it to provide some context. It does explain later that "vowel" will refer to each character, but it might be worth mentioning that you can access string characters by index as well before jumping into this loop. Alternatively, removing the iteration over strings altogether might be a good idea as well so as to not introduce too many new ideas at once.

**Comment by maxim-belkin** [comment]
Thanks, Kristen. I agree that jumping from iterating over lists to iterating over strings might be a bit odd (heh). If you'd like to propose any changes with a pull request -- that would be very much appreciated.

@kekoziar, apologies for not replying to your question but in order to answer it I have to re-read the first 3 episodes together and I wasn't able to find time for this exercise yet...

**Comment by kekoziar** [comment]
I also welcome any clarifications for that section. I agree it's a little clunky.  While strings can be read using index notation, they can't be accessed using read/write like lists.  I usually note when teaching the loop examples that use strings that while characters in strings can be read using index notation, there are other, faster, more powerful methods for working with strings.

**Comment by biokcb** [comment]
@kekoziar Yeah, explaining that strings can be treated similarly to lists in some ways, but not all might be too confusing at this stage. I'm leaning toward just not introducing this string iteration concept at all, thoughts?

I can certainly make some proposed changes via a pull request! I can work on this sometime tomorrow and will tag you both for feedback if that's ok.

**Comment by bhickson** [comment]
I'm good with removing it. Rarely actually so string slicing myself, but
it does illuminate a bit about how they work; which I suppose is why I put
it in in the first place.

**Comment by ewinge** [comment]
Is anyone working on this issue? In my opinion, this should be a priority. Iterating over the characters in a string is too obscure for an introductory course.
Unfortunately, the lists-loops branch has fallen behind.

**Comment by ldko** [comment]
Hi Erik, as far as I know, no one is actively working on this. If you would like to take it up (i.e. getting the branch up-to-date, resolving any conflicts, checking for references that no long make sense, etc.), please do! Thank you!

**Comment by ewinge** [comment]
Hi, I have made PR #895 please have a look

**Comment by biokcb** [comment]
Hi @ewinge - thanks for pushing these changes up! I hadn't kept up with the progress of my changes in the list-loops branch being taken to main. Out of curiosity are you pulling in the list-loops branch or did you just re-write it?

**Comment by biokcb** [comment]
For context @ewinge @maxim-belkin: I was waiting for someone to want to merge my changes into the main branch (seems to have fallen off the radar and now the branch is out-of-date), we worked in the list-loops branch since it was a major change. Would you mind if we tried to update the list-loops branch instead as @ldko mentioned? I'm kind of sad now that my changes won't get merged and was busy at the time this first got pinged again...

**Comment by ewinge** [comment]
Hi @biokcb, the list-loops branch had fallen behind, so it was easier to start from scratch.

**Comment by biokcb** [comment]
Hi @ewinge I agree it would be nice to have the lesson finally implement this change in the `gh-pages` branch. I didn't realize it wasn't til I saw this email. 

>I don’t really care who’s PR gets merged. However, this issue will soon have been open for three years.
Isn’t it time to get it resolved? 

I feel like this comment is a bit hostile and unnecessary. I only mentioned my feelings on it for context behind my request for it since I don't have an official role on the repository. In the original issue statement from @maxim-belkin it specifically asks to work in the `list-loops` branch & make any pull requests against that branch, where others made PRs that were merged. I mentioned the context to reiterate this point and imagine that's why @ldko mentioned it too, but I am not trying to enforce anything. I just wanted to ask in case we could update it this way or just if you'd even consider it. I apologize if it came off as an annoying request, but no need to be rude about it. I understand it is easier to work in `gh-pages` due to the number of updates that have not been kept in `list-loops` and it would have taken a little more time to update the branch so a simple "no, I think it is quicker/better this way because of the number of commits to merge" would have sufficed.

**Comment by ewinge** [comment]
Hi @biokcb, I apologize if I was rude, that was not my intention:)

**Comment by maxim-belkin** [comment]
Hi @biokcb and @ewinge.
Yes, it did fell off my radar for a number of reasons -- my apologies. Let's proceed with the fastest approach, which I think is to:

1. Update the `lists-loops` branch with the changes from `gh-pages` (this might require a rebase or a merge). This is necessary because we split the first episode into three about a year ago. This will preserve all contributions to the `lists-loops` branch.
2. (In case of a rebase in step 1): Force-push the updated `lists-loops` branch. Because of this, all contributors who have previously checked out the `lists-loops` branch locally would have to hard-reset their local `lists-loops` branch like so:
   ```
   git fetch origin lists-loops 
   git checkout lists-loops
   git reset --hard origin/lists-loops
   ```
3. After that we can review the changes remaining in #496 that we still apply. But as @biokcb said, it has to be submitted against the `lists-loops` branch because we'd have to announce this change on our mailing list before merging it into `gh-pages`.

Thanks all for your patience!

**Comment by maxim-belkin** [comment]
I've rebased the `lists-loops` branch on top of `gh-pages` and opened a pull request (#926). We need to double check everything and, when things look good, announce the change on The Carpentries' Discuss channel. Right now I see one issue there -- the sentence that begins with `In the first episode, we wrote Python code that` has to be updated to reflect the fact that the first episode was split into three.
